tiny_dnn/util/product.h

```diff
@@ -217,218 +257,357 @@ struct double_avx {
+    auto s = T::template load<src_aligned>(&src[idx + i * sz]);
+    d      = T::add(s, d);
+    T::template store<dst_aligned>(&dst[idx + i * sz], d);
+  }
+  idx += n1 * sz;
+  for (size_t i = 0; i < remain; ++i) {
+    dst[idx + i] += src[idx + i];
   }
-
-  for (std::size_t i = (size / T::unroll_size) * T::unroll_size; i < size; i++)
-    dst[i] += src[i] * c;
 }
 
-template <typename T>
-inline void reduce_nonaligned(const typename T::value_type *src,
-                              std::size_t size,
-                              typename T::value_type *dst) {
-  for (std::size_t i = 0; i < size / T::unroll_size; i++) {
-    typename T::register_type d = T::loadu(&dst[i * T::unroll_size]);
-    typename T::register_type s = T::loadu(&src[i * T::unroll_size]);
-    T::storeu(&dst[i * T::unroll_size], T::add(d, s));
+template <typename T, typename dst_aligned>
+inline void fill_impl(typename T::value_type *dst,
+                      std::size_t size,
+                      typename T::value_type value) {
+  auto vals   = T::set1(value);
+  auto sz     = T::unroll_size;
+  auto sz4    = T::unroll_size * 4;
+  auto n4     = size / sz4;
+  auto n1     = (size % sz4) / sz;
+  auto remain = size % sz;
+  for (size_t i = 0; i < n4; ++i) {
+    T::template store<dst_aligned>(&dst[i * sz4 + sz * 0], vals);
+    T::template store<dst_aligned>(&dst[i * sz4 + sz * 1], vals);
+    T::template store<dst_aligned>(&dst[i * sz4 + sz * 2], vals);
+    T::template store<dst_aligned>(&dst[i * sz4 + sz * 3], vals);
+  }
+  size_t idx = n4 * sz4;
+  for (size_t i = 0; i < n1; ++i) {
+    T::template store<dst_aligned>(&dst[idx + i * sz], vals);
+  }
+  idx += n1 * sz;
+  for (size_t i = 0; i < remain; ++i) {
+    dst[idx + i] += value;
   }
-
-  for (std::size_t i = (size / T::unroll_size) * T::unroll_size; i < size; i++)
-    dst[i] += src[i];
 }
 
+#if defined(CNN_USE_AVX)
+#ifdef CNN_USE_DOUBLE
+#define VECTORIZE_TYPE detail::double_avx
+#else
+#define VECTORIZE_TYPE detail::float_avx
+#endif
+#elif defined(CNN_USE_SSE)
+#ifdef CNN_USE_DOUBLE
+#define VECTORIZE_TYPE detail::double_sse
+#else
+#define VECTORIZE_TYPE detail::float_sse
+#endif
+#else
+#ifdef CNN_USE_DOUBLE
+#define VECTORIZE_TYPE detail::scalar_generic<double>
+#else
+#define VECTORIZE_TYPE detail::scalar_generic<float>
+#endif
+#endif
+
 template <typename T>
-inline void reduce_aligned(const typename T::value_type *src,
-                           std::size_t size,
-                           typename T::value_type *dst) {
-  for (std::size_t i = 0; i < size / T::unroll_size; i++) {
-    typename T::register_type d = T::load(&dst[i * T::unroll_size]);
-    typename T::register_type s = T::load(&src[i * T::unroll_size]);
-    T::store(&dst[i * T::unroll_size], T::add(d, s));
+void fill(T *dst, size_t size, T value) {
+// On recent x86/x64 processors, stos instructions are faster
+// than SSE/AVX store instructions and recent compilers generates
+// stos instructions with std::fill
+#if 1
```

Comment:
This comment explains why simple `std::fill` is used over inhouse `vectorize::detail::fill_impl`.
Simple unrolled SSE/AVX store instructions isn't the fastest with recent x86/x64 processors.

With Visual Studio, there're intrinsic functions for `stosd` instruction but I can't find equivalent once for gcc/clang. Fortunately, compilers follow `64-ia-32-architectures-optimization-manual.pdf` so they produce right instructions.